src/main/java/net/machpi/runelite/influxdb/activity/ActivityState.java

```diff
@@ -0,0 +1,136 @@
+package net.machpi.runelite.influxdb.activity;
+
+import com.google.common.collect.ComparisonChain;
+import lombok.Data;
+import lombok.extern.slf4j.Slf4j;
+import net.machpi.runelite.influxdb.InfluxDbConfig;
+import net.machpi.runelite.influxdb.MeasurementCreator;
+import net.machpi.runelite.influxdb.write.InfluxWriter;
+import net.machpi.runelite.influxdb.write.Measurement;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+@Slf4j
+public class ActivityState {
+
+    @Data
+    private static class EventWithTime {
+        private final GameEvent type;
+        private final Instant start;
+        private Instant updated;
+    }
+
+    @Data
+    public static class State {
+        private final String skill;
+        private final String location;
+        private final String locationType;
+    }
+
+    private final List<EventWithTime> events = new ArrayList<>();
+    private final InfluxDbConfig config;
+    private final MeasurementCreator measurer;
+    private final InfluxWriter writer;
+    private State lastState;
+
+    public ActivityState(final InfluxDbConfig config, final InfluxWriter writer, final MeasurementCreator measurer) {
+        this.config = config;
+        this.writer = writer;
+        this.measurer = measurer;
+    }
+
+    /**
+     * Reset state.
+     */
+    public void reset() {
+        events.clear();
+        lastState = null;
+    }
+
+    /**
+     * Trigger new state update.
+     *
+     * @param eventType event type
+     */
+    public void triggerEvent(final GameEvent eventType) {
+        if (!config.writeActivity()) return;
+
+        final Optional<EventWithTime> foundEvent = events.stream().filter(e -> e.type == eventType).findFirst();
```

Comment:
Not a fan of unnecessary iteration.  You could instead be maintaining a `Map<EventType, EventData>` and a `TreeSet<EventData>`.